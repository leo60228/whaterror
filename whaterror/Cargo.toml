[package]
name = "whaterror"
version = "0.1.0"
authors = ["leo60228 <leo@60228.dev>"]
edition = "2018"
license = "MIT OR Apache-2.0"
description = "Customizable handling for fallible main"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
