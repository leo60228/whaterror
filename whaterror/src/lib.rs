use std::convert::Infallible;

pub trait FatalError<T> {
    fn handle(self, handler: T);
}

pub trait Termination<T> {
    type Err: FatalError<T>;

    fn into_result(self) -> Result<(), Self::Err>;
}

impl<T> FatalError<T> for Infallible {
    fn handle(self, _handler: T) {
        match self {}
    }
}

impl<T> Termination<T> for () {
    type Err = Infallible;

    fn into_result(self) -> Result<(), Self::Err> {
        Ok(self)
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
    }
}
