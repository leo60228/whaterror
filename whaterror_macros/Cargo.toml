[package]
name = "whaterror_macros"
version = "0.1.0"
authors = ["leo60228 <leo@60228.dev>"]
edition = "2018"
license = "MIT OR Apache-2.0"
description = "Macros for whaterror"

[lib]
proc-macro = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
proc-macro-error = "1.0.4"
proc-macro2 = "1.0.24"
quote = "1.0.8"
syn = { version = "1.0.58", features = ["full"] }

[dev-dependencies]
trybuild = "1.0.38"
whaterror = { path = "../whaterror" }
