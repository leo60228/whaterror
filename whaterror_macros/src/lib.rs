use proc_macro::TokenStream as ProcTokenStream;
use proc_macro2::{Span, TokenStream};
use proc_macro_error::*;
use quote::ToTokens;
use syn::{
    parse::{Parse, ParseStream, Result},
    parse_quote,
    spanned::Spanned,
    Expr, ExprBlock, ExprClosure, Ident, ItemFn, ReturnType, TypeTuple,
};

fn try_fold1<I, F>(mut iter: I, f: F) -> Option<I::Item>
where
    I: Iterator,
    F: FnMut(I::Item, I::Item) -> Option<I::Item>,
{
    let elem = iter.next()?;
    iter.try_fold(elem, f)
}

struct MacroArgs {
    expr: Expr,
}

impl Parse for MacroArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let expr = Expr::parse(input)?;

        Ok(Self { expr })
    }
}

#[proc_macro_error]
#[proc_macro_attribute]
pub fn whaterror(attr: ProcTokenStream, item: ProcTokenStream) -> ProcTokenStream {
    dummy::set_dummy(item.clone().into());

    let attr: TokenStream = attr.into();
    let item: TokenStream = item.into();

    let args = if attr.is_empty() {
        emit_call_site_error!("missing arguments");
        None
    } else {
        let span = SpanRange::from_tokens(&attr);

        match syn::parse2::<MacroArgs>(attr) {
            Ok(args) => Some(args),
            Err(err) => {
                emit_error!(span, "{}", err);
                None
            }
        }
    };

    let inner_main_fn = syn::parse2::<ItemFn>(item).unwrap_or_abort();

    if let Some(constness) = &inner_main_fn.sig.constness {
        emit_error!(constness.span(), "const fns are not supported");
    }

    if let Some(asyncness) = &inner_main_fn.sig.asyncness {
        emit_error!(asyncness.span(), "async fns are not supported");
    }

    if !inner_main_fn.sig.generics.params.is_empty()
        || inner_main_fn.sig.generics.where_clause.is_some()
    {
        emit_error!(
            inner_main_fn.sig.generics.span(),
            "generic functions are not supported"
        );
    }

    if !inner_main_fn.sig.inputs.is_empty() {
        emit_error!(
            inner_main_fn.sig.inputs.span(),
            "arguments are not supported"
        );
    }

    if !inner_main_fn.attrs.is_empty() {
        emit_warning!(
            try_fold1(inner_main_fn.attrs.iter().map(Spanned::span), |a, b| a
                .join(b))
            .unwrap_or_else(Span::call_site),
            "attributes may have unexpected behavior"
        );
    }

    abort_if_dirty();

    let args = args.unwrap();
    let expr = args.expr;

    let mut outer_main = inner_main_fn.clone();

    let inner_main = ExprClosure {
        attrs: vec![],
        asyncness: None,
        movability: None,
        capture: Some(Default::default()),
        or1_token: Default::default(),
        inputs: Default::default(),
        or2_token: Default::default(),
        output: match inner_main_fn.sig.output {
            ReturnType::Default => ReturnType::Type(
                Default::default(),
                Box::new(
                    TypeTuple {
                        paren_token: Default::default(),
                        elems: Default::default(),
                    }
                    .into(),
                ),
            ),
            x @ ReturnType::Type(_, _) => x,
        },
        body: Box::new(
            ExprBlock {
                attrs: vec![],
                label: None,
                block: *inner_main_fn.block,
            }
            .into(),
        ),
    };

    let inner = Ident::new("inner", Span::mixed_site());
    let thunk = Ident::new("thunk", Span::mixed_site());

    outer_main.sig.output = ReturnType::Default;
    outer_main.block = Box::new(parse_quote! {{
        let #inner = #inner_main;
        let #thunk = || #expr;

        {
            use ::whaterror::{Termination, FatalError, terminate};

            // help out type inference
            fn handle<R: Termination<H>, H>(result: R, handler: fn() -> H) {
                if let Err(e) = result.into_result() {
                    e.handle(handler());
                    terminate(cfg!(test));
                }
            }

            handle(#inner(), #thunk);
        }
    }});

    outer_main.into_token_stream().into()
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
    }
}
