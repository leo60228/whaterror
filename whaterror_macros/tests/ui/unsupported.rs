#![feature(c_variadic)]

use whaterror_macros::whaterror;

#[whaterror(())]
const fn const_fn() {}

#[whaterror(())]
async fn async_fn() {}

#[whaterror(())]
fn generic_fn<T>() {}

#[whaterror(())]
fn arguments(_x: i32) {}

#[whaterror(())]
unsafe extern "C" fn variadic(_arg: i32, _args: ...) {}

#[whaterror(())]
#[warn(warnings)]
fn attrs() {}

#[whaterror(())]
async fn generic_async_args<T>(_x: T) {}

fn main() {}
