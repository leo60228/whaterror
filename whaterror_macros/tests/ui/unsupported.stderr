error: const fns are not supported
 --> $DIR/unsupported.rs:6:1
  |
6 | const fn const_fn() {}
  | ^^^^^

error: async fns are not supported
 --> $DIR/unsupported.rs:9:1
  |
9 | async fn async_fn() {}
  | ^^^^^

error: generic functions are not supported
  --> $DIR/unsupported.rs:12:14
   |
12 | fn generic_fn<T>() {}
   |              ^^^

error: arguments are not supported
  --> $DIR/unsupported.rs:15:14
   |
15 | fn arguments(_x: i32) {}
   |              ^^^^^^^

error: arguments are not supported
  --> $DIR/unsupported.rs:18:31
   |
18 | unsafe extern "C" fn variadic(_arg: i32, _args: ...) {}
   |                               ^^^^^^^^^^^^^^^^^^^^^

warning: attributes may have unexpected behavior
  --> $DIR/unsupported.rs:21:1
   |
21 | #[warn(warnings)]
   | ^^^^^^^^^^^^^^^^^

error: async fns are not supported
  --> $DIR/unsupported.rs:25:1
   |
25 | async fn generic_async_args<T>(_x: T) {}
   | ^^^^^

error: generic functions are not supported
  --> $DIR/unsupported.rs:25:28
   |
25 | async fn generic_async_args<T>(_x: T) {}
   |                            ^^^

error: arguments are not supported
  --> $DIR/unsupported.rs:25:32
   |
25 | async fn generic_async_args<T>(_x: T) {}
   |                                ^^^^^
